std_execution.bs

```diff
@@ -4555,6 +4567,71 @@ from the operation before the operation completes.
 
         - Otherwise, <code>execution::set_done(<i>GET-BASE</i>(static_cast&lt;Derived&amp;&amp;>(self)))</code>
 
+### `execution::receiver_signatures` <b>[execution.snd_rec_utils.rcvr_sigs]</b> ### {#spec-execution.snd_rec_utils.rcvr_sigs}
+
+1. `receiver_signatures` is used to define a type that implements the nested `value_types`, `error_types`, and `sends_done` members that describe a typed sender. Its arguments are a flat list of function types that describe the signatures of the receiver's completion-signal operations that the sender invokes.
+
+2. [<i>Example:</i>
+     <pre highlight="c++">
+     using my_traits =
+       execution::receiver_signatures&lt;
+         execution::set_value_t(),
+         execution::set_value_t(int, float),
+         execution::set_error_t(exception_ptr),
+         execution::set_error_t(error_code),
+         execution::set_done_t()
+       >;
+
+      // my_traits::value_types&lt;tuple, variant> names the type:
+      //   variant&lt;tuple&lt;>, tuple&lt;int, float>>
+      //
+      // my_traits::error_types&lt;variant> names the type:
+      //   variant&lt;exception_ptr, error_code>
+      //
+      // my_traits::sends_done is true
+     </pre>
+     -- <i>end example</i>]
+
+3. This section makes use of the following exposition-only concept:
+
+    <pre highlight="c++">
+    template &lt;class Fn>
+      concept <i>receiver-signature</i> = <i>see below</i>;
+    </pre>
+
+    1. A type `Fn` satisfies <code><i>receiver-signature</i></code> if it is a function type with one of the following forms:
+
+        * <code>set_value_t(<i>Vs</i>...)</code>, where <code><i>Vs</i></code> is an arbitrary parameter pack.
+        * <code>set_error_t(<i>E</i>)</code>, where <code><i>E</i></code> is an arbitrary type.
+        * `set_done_t()`
+
+    2. Otherwise, `Fn` does not satisfy <code><i>receiver-signature</i></code>.
+
+4.  <pre highlight="c++">
+    template &lt;<i>receiver-signature</i>... Fns>
+      using receiver_signagures = <i>see below</i>;
+    </pre>
+
+    Given a template parameter pack <code><i>Fns</i></code> of types that satisfy <code><i>receiver-signature</i></code>, <code>receiver_signatures&lt;<i>Fns</i>...></code> names a non-template class type equivalent to the following:
```

Comment:
> ... names a non-template class type ...

I will try to write up a paper proposing your "private" template parameters for class templates as a way of giving better control of ADL so we can avoid having to jump through hoops like this.

Reply:
> > ... names a non-template class type ...
> 
> I will try to write up a paper proposing your "private" template parameters for class templates as a way of giving better control of ADL so we can avoid having to jump through hoops like this.

That's nice, but please refrain from using 'private' in that even as a stopgap working todo-spelling of the facility. It's a colossal distraction, considering that name lookup and access control are more or less completely orthogonal in C++.

Reply:
Obviously, you're right, `private` is for controlling access, and it's helpful to know what to expect of we bring this to EWG.

Access control is, in part, for marking parts of a class as implementation details. And this is another time you want to mark a part of a class as an implementation detail. That's why I think `private` isn't the worst option.

But I'm curious if you have any other suggestions.